### ISSUE TITLE
Bogus error "type of variable template specialization [...] declared with deduced type [...] depends on itself"

### ISSUE THREAD

**Issue by mattkretz** [llvm/llvm-project#214477]
Test case:

```c++
#include <type_traits>

struct dummy {
};

template <typename A> struct real { A a; };

template <typename V> struct Tests {
  template <int> static constexpr dummy obj = {};

  const int reg = [] {
    using X = decltype(obj<0>);  // #1
    (void)&obj<0>;               // #2
    return 0;
  }();

  template <int Tmp>
    requires(Tmp == 0)
  static constexpr real obj<Tmp> = {0};
};

int main() { Tests<int>{}; }
```

```sh
❯ clang++-23 -std=c++26 -O2 -Wall clang_issue.cpp
clang_issue.cpp:13:12: error: the type of variable template specialization 'obj<0>' declared with deduced type 'const real'
      depends on itself
   13 |     (void)&obj<0>;
      |            ^
clang_issue.cpp:11:22: note: while substituting into a lambda expression here
   11 |   const int reg = [] {
      |                      ^
clang_issue.cpp:22:25: note: in instantiation of default member initializer 'Tests<int>::reg' requested here
   22 | int main() { Tests<int>{}; }
      |                         ^
clang_issue.cpp:22:14: warning: expression result unused [-Wunused-value]
   22 | int main() { Tests<int>{}; }
      |              ^         ~~
1 warning and 1 error generated.

```

Comment line `#1` or swap lines `#1` and `#2` and the code is accepted. GCC always accepts.

The error message is not helpful. But I very much suspect Clang is incorrect to reject this program.

**Comment by zwuis** [comment]
Related: #178860.

**Comment by shafik** [comment]
CC @Endilll

**Comment by mattkretz** [comment]
Observation: instead of `using X = decltype(obj<0>);` declare a variable of type `decltype(obj<0>)` and the error message suggests that Clang tries to create an object of type `real`. I.e. the diagnostics make it sound like it tries to use a class template instead of an specialization of the template.
